### ISSUE TITLE
Bug in Editor: Some STL models appears in black

### ISSUE THREAD

**Issue by lrusso** [mrdoob/three.js#12499]
Some STL models appears in black in the editor (https://threejs.org/editor), for example: https://www.thingiverse.com/thing:45177

The following two lines of code are required to fix this at the moment to load the model:

geometry.computeFaceNormals();
geometry.computeVertexNormals();

**Comment by WestLangley** [comment]
That's a problem with the model -- not the editor. The model's normals are all zero.

**Comment by WestLangley** [comment]
The model's normals are all zero. So either the model is in error, or the loader has a bug.

**Comment by lrusso** [comment]
After converting the following 3DS model from the Three.js Web (https://threejs.org/examples/models/3ds/portalgun/portalgun.3ds) into a STL model, i got the same issue, fixed with those two lines of code.

![three js editor](https://user-images.githubusercontent.com/11498901/32033083-e47c296a-b9e0-11e7-897c-6f386e9f9d7a.png)
![three js fixed](https://user-images.githubusercontent.com/11498901/32033084-e4a9278a-b9e0-11e7-9186-723a606829ce.png)

**Comment by Mugen87** [comment]
Maybe it's not relevant but you need to add a light source to your scene. Otherwise the model remains black. The editor applies per default an instance of `MeshStandardMaterial` to a `Mesh` that uses the geometry of the STL file.

So when i load this [file](https://github.com/mrdoob/three.js/blob/dev/examples/models/stl/ascii/slotted_disk.stl) in the editor and add a light source, the result looks good.

![image](https://user-images.githubusercontent.com/12612165/32042331-741fee70-ba37-11e7-8ddd-db3929408586.png)

**Comment by mrdoob** [comment]
Sometimes I wonder if we should add a default lighting to the scene if the user hasn't added lights. Or maybe a default `HemisphereLight` in the editor?

**Comment by lrusso** [comment]
When adding a light source, some STL objects (The portalgun.stl and the Trumpet Mouthpiece) are not affected by it. It works fine with the other STL object (the Einstein model). This can be fixed by adding those two lines of code.

![three js_editor_-_2017-10-26_16 14 59](https://user-images.githubusercontent.com/11498901/32072409-da174a4e-ba68-11e7-8531-ca7d14ac90a0.png)

![model](https://user-images.githubusercontent.com/11498901/32072752-1f7e9e88-ba6a-11e7-9eea-d0bea5a6c507.png)

**Comment by Mugen87** [comment]
> Sometimes I wonder if we should add a default lighting to the scene if the user hasn't added lights. 

I like the idea. Unity for example provides a directional light in new projects. 

![image](https://user-images.githubusercontent.com/12612165/32073526-adfabfee-ba96-11e7-811b-963d55ae3cc2.png)

**Comment by Mugen87** [comment]
@lrusso I don't think this is a problem by the loader. Users have to ensure that their `STL` files have valid normal data. If they don't have these information they can try to generate them with `geometry.computeVertexNormals()`. But from my point of view, it's not a task of the loader.

**Comment by WestLangley** [comment]
https://github.com/mrdoob/three.js/issues/12499#issuecomment-339534784

It should be easy enough to determine if the loader is correctly loading the normals as specified in the model.

@mrdoob I would suggest this pattern if you want to add default lighting to the editor:
```javascript
scene.add( camera );
camera.add( pointLight );
```

**Comment by mrdoob** [comment]
> > Sometimes I wonder if we should add a default lighting to the scene if the user hasn't added lights. 
>
> I like the idea. Unity for example provides a directional light in new projects. 
>
> ![image](https://user-images.githubusercontent.com/12612165/32073526-adfabfee-ba96-11e7-811b-963d55ae3cc2.png)

Good to know! Yeah, I think this makes sense. I find myself adding a `DirectionalLight` every single time anyway.

> @mrdoob I would suggest this pattern if you want to add default lighting to the editor:
> ```javascript
> scene.add( camera );
> camera.add( pointLight );
> ```

I think that would be a bit confusing for the user...

**Comment by Drenmi** [comment]
+1 for a default light source. 🙂

**Comment by Mugen87** [comment]
Closing since `STLLoader` does its job correctly. Adding a default lighting to the editor will be considered separately.

**Comment by cowlicks** [comment]
@Mugen87 note that according to Wikipedia: ["In most software this may be set to (0,0,0), and the software will automatically calculate a norma"](https://en.wikipedia.org/wiki/STL_%28file_format%29#The_facet_normal).

So the STLLoader should probably auto-calculate the normal if they are set to zero

**Comment by mrdoob** [comment]
@cowlicks That seems fair. Would you like to do a PR with the fix?

**Comment by WestLangley** [comment]
Hmm... I think the loader should just load the file -- not change the data.

An exception to that convention would be if the STL spec specified that normals were optional -- or that zero normals were valid. I have not been able to find any STL specification that says that.

The workaround is for the user to call `.computeVertexNormals()` in the loader callback.

**Comment by mrdoob** [comment]
This issue is about the Editor. I guess the editor should call `. computeVertexNormals()`?
